File: src/gosim/game/war.go
package game

import (
	"math/rand"
)

// WarGame represents the game state for War
type WarGame struct {
	Player1Hand []int
	Player2Hand []int
	Turns       int
	rng         *rand.Rand
}

// WarResult contains game outcome
type WarResult struct {
	Winner int
	Turns  int
}

// NewWarGame creates a new War game
func NewWarGame(seed int64) *WarGame {
	rng := rand.New(rand.NewSource(seed))

	// Create deck (ranks 1-13, four of each)
	deck := make([]int, 52)
	idx := 0
	for suit := 0; suit < 4; suit++ {
		for rank := 1; rank <= 13; rank++ {
			deck[idx] = rank
			idx++
		}
	}

	// Shuffle
	rng.Shuffle(len(deck), func(i, j int) {
		deck[i], deck[j] = deck[j], deck[i]
	})

	return &WarGame{
		Player1Hand: deck[:26],
		Player2Hand: deck[26:],
		Turns:       0,
		rng:         rng,
	}
}

// PlayBattle plays one battle
func (g *WarGame) PlayBattle() {
	if len(g.Player1Hand) == 0 || len(g.Player2Hand) == 0 {
		return
	}

	p1Card := g.Player1Hand[0]
	p2Card := g.Player2Hand[0]
	g.Player1Hand = g.Player1Hand[1:]
	g.Player2Hand = g.Player2Hand[1:]

	if p1Card > p2Card {
		g.Player1Hand = append(g.Player1Hand, p1Card, p2Card)
	} else if p2Card > p1Card {
		g.Player2Hand = append(g.Player2Hand, p2Card, p1Card)
	} else {
		// War!
		if len(g.Player1Hand) >= 4 && len(g.Player2Hand) >= 4 {
			warPile := []int{p1Card, p2Card}
			warPile = append(warPile, g.Player1Hand[:4]...)
			warPile = append(warPile, g.Player2Hand[:4]...)
			g.Player1Hand = g.Player1Hand[4:]
			g.Player2Hand = g.Player2Hand[4:]

			// Winner takes all
			if warPile[len(warPile)-4] > warPile[len(warPile)-1] {
				g.Player1Hand = append(g.Player1Hand, warPile...)
			} else {
				g.Player2Hand = append(g.Player2Hand, warPile...)
			}
		} else {
			// Not enough cards, return them
			g.Player1Hand = append(g.Player1Hand, p1Card)
			g.Player2Hand = append(g.Player2Hand, p2Card)
		}
	}

	g.Turns++
}

// IsGameOver checks if game has ended
func (g *WarGame) IsGameOver() bool {
	return len(g.Player1Hand) == 0 || len(g.Player2Hand) == 0
}

// GetWinner returns winner (1 or 2)
func (g *WarGame) GetWinner() int {
	if len(g.Player1Hand) > len(g.Player2Hand) {
		return 1
	}
	return 2
}

// PlayWarGame plays a complete game
func PlayWarGame(seed int64, maxTurns int) WarResult {
	game := NewWarGame(seed)

	for !game.IsGameOver() && game.Turns < maxTurns {
		game.PlayBattle()
	}

	return WarResult{
		Winner: game.GetWinner(),
		Turns:  game.Turns,
	}
}

File: src/darwindeck/simulation/war.py
"""War game simulation (Python baseline)."""

import random
from typing import Dict, List, Tuple


class WarGame:
    """Simple War card game implementation."""

    def __init__(self, seed: int = 42) -> None:
        """Initialize War game with shuffled deck."""
        self.rng = random.Random(seed)

        # Create deck (rank only matters for War)
        deck = list(range(1, 14)) * 4  # 1-13, four suits
        self.rng.shuffle(deck)

        # Split evenly
        self.player1_hand = deck[:26]
        self.player2_hand = deck[26:]
        self.turns = 0

    def play_battle(self) -> None:
        """Play one battle."""
        if not self.player1_hand or not self.player2_hand:
            return

        p1_card = self.player1_hand.pop(0)
        p2_card = self.player2_hand.pop(0)

        if p1_card > p2_card:
            self.player1_hand.extend([p1_card, p2_card])
        elif p2_card > p1_card:
            self.player2_hand.extend([p2_card, p1_card])
        else:
            # War! Each player plays 3 face down + 1 face up
            if len(self.player1_hand) >= 4 and len(self.player2_hand) >= 4:
                war_pile = [p1_card, p2_card]
                war_pile.extend(self.player1_hand[:4])
                war_pile.extend(self.player2_hand[:4])
                self.player1_hand = self.player1_hand[4:]
                self.player2_hand = self.player2_hand[4:]

                # Winner takes all
                if war_pile[-4] > war_pile[-1]:  # p1 wins
                    self.player1_hand.extend(war_pile)
                else:  # p2 wins
                    self.player2_hand.extend(war_pile)
            else:
                # Not enough cards for war, return cards
                self.player1_hand.append(p1_card)
                self.player2_hand.append(p2_card)

        self.turns += 1

    def is_game_over(self) -> bool:
        """Check if game has ended."""
        return len(self.player1_hand) == 0 or len(self.player2_hand) == 0

    def get_winner(self) -> int:
        """Get winner (1 or 2)."""
        if len(self.player1_hand) > len(self.player2_hand):
            return 1
        return 2


def play_war_game(seed: int = 42, max_turns: int = 1000) -> Dict[str, int]:
    """Play a complete War game and return results."""
    game = WarGame(seed=seed)

    while not game.is_game_over() and game.turns < max_turns:
        game.play_battle()

    return {
        "winner": game.get_winner(),
        "turns": game.turns
    }
